File: apps/android-receiver/app/src/main/java/com/antigravity/opticalreceiver/codec/CRC32C.kt
package com.antigravity.opticalreceiver.codec

object CRC32C {
    private val table = IntArray(256)

    init {
        for (i in 0 until 256) {
            var c = i
            for (k in 0 until 8) {
                c = if ((c and 1) != 0) (0x82F63B78.toInt() xor (c ushr 1)) else (c ushr 1)
            }
            table[i] = c
        }
    }

    fun calculate(data: ByteArray, seed: Int = 0): Long {
        var crc = seed xor 0xFFFFFFFF.toInt()
        for (b in data) {
            val byteVal = b.toInt() and 0xFF
            crc = (crc ushr 8) xor table[(crc xor byteVal) and 0xFF]
        }
        return (crc xor 0xFFFFFFFF.toInt()).toLong() and 0xFFFFFFFFL
    }
}

File: packages/codec/src/crc32c.ts
/**
 * Hardware-inspired lookup table implementation of CRC32C (Castagnoli)
 * Polynomial: 0x82F63B78
 */

const CRC32C_TABLE = new Uint32Array(256);

for (let i = 0; i < 256; i++) {
  let c = i;
  for (let k = 0; k < 8; k++) {
    c = (c & 1) ? (0x82F63B78 ^ (c >>> 1)) : (c >>> 1);
  }
  CRC32C_TABLE[i] = c >>> 0;
}

export function crc32c(data: Uint8Array, seed: number = 0): number {
  let crc = seed ^ 0xFFFFFFFF;
  for (let i = 0; i < data.length; i++) {
    crc = (crc >>> 8) ^ CRC32C_TABLE[(crc ^ data[i]) & 0xFF];
  }
  return (crc ^ 0xFFFFFFFF) >>> 0;
}
